### ISSUE TITLE
Deepwave for undulating surface

### ISSUE THREAD

**Issue by ruihua-chen** [ar4/deepwave#95]
Could you please tell me if Deepwave supports the setting of free boundaries for undulating surface? If it doesn't support this, are there any ways or methods to modify it?

**Comment by CRuihua** [comment]
I attempted to modify the finite difference part of deepwave, trying to set the gradient conditions for free boundaries on the undulating surface. However, it seems that "forward" points to the .so file instead of the python file.

**Comment by ar4** [comment]
Dear @CRuihua 

Thank you for your question. I am currently working on the next release of Deepwave, in which my current plan is to switch the elastic propagator to use the improved vacuum formulation of Zeng et al. One advantage of this approach is that the free surface can be specified simply by setting the model parameters above the surface to zero. This means that users can easily have topography and even voids in their model. I would prefer to use this type of approach, where topography is handled automatically rather than requiring special treatment. I am not sure whether a similar approach is possible with the scalar wave propagator.

There is still quite a lot of work to do before the next release, so I expect that it will be a few months before it is ready. If you wish to make a modified version of the propagator, you will need to edit the C or CUDA code. You can download the code from GitHub (https://github.com/ar4/deepwave/archive/refs/heads/master.zip) and then you will find the C and CUDA files in `src/deepwave`. Once you have made the changes you want, I will be happy to help you to compile and use your modified version.

**Comment by ar4** [comment]
That is correct. The computationally intensive parts are written in C (for CPUs) and CUDA (for GPUs). Both the forward and backward are implemented this way (for example, [here is the code for backpropagation for scalar propagation on a CPU](https://github.com/ar4/deepwave/blob/00798f86eb96c80294b8aed69569bab63417041e/src/deepwave/scalar.c#L117)). It therefore manually calculates the gradients rather than using PyTorch's automatic differentiation, but it does this in a way that integrates with PyTorch, so that end-to-end gradient calculation still works with any operations that you apply to the inputs before passing them to Deepwave, and anything you do to the outputs of Deepwave.

It is possible to implement wave propagation purely using PyTorch, and let PyTorch do automatic differentiation. I am planning to add this as an option in the next release. It is, however, quite a lot slower than the C/CUDA code.

**Comment by CRuihua** [comment]
Following your valuable suggestion,  I will attempt to configure the immersed boundary method in the C language environment of Deepwave to achieve forward modeling of the acoustic wave equation for irregular topography!

**Comment by ar4** [comment]
The performance of PyTorch continues to improve, so I hope that one day it might be possible to get adequate performance with a pure PyTorch implementation of wave propagation, but that is unfortunately not yet the case.

Please let me know if you encounter any difficulties or have any questions while implementing your change in C.

**Comment by CRuihua** [comment]
Thank you very much for your reply. It has given me great confidence. By the way, do I need to amend the backward part after implementing the immersed fluctuating surface free boundary condition?

**Comment by ar4** [comment]
Hi Ruihua,

Do I correctly understand that you modified [the example that uses checkpointing to reduce memory requirements for inversion](https://github.com/ar4/deepwave/blob/master/docs/example_checkpointing.py), using regular Deepwave (not the modification you are working on to implement a free surface with topography), but the result of forward propagation was not the same as if you had not used checkpointing? If so, you are right that that should not occur. Would it be possible for you to share your code (preferably a simple example that demonstrates the problem) so that I can try to determine the cause?

**Comment by AI4FWI** [comment]
Hi Alan, I also conducted tests on the Marmousi model and found that there were indeed problems. When I set the checkpoint point to 1, no artifacts would occur. However, if I set it to a larger value, a corresponding number of artifacts would appear. I'm not sure if it's a problem with my code or with the DeepWave itself. Here are my code:

<img width="520" height="345" alt="Image" src="https://github.com/user-attachments/assets/266eac9d-7f0b-4548-8195-2087720dddf3" />

import torch
import matplotlib.pyplot as plt
import deepwave
from deepwave import scalar

device = torch.device('cuda' if torch.cuda.is_available()
                      else 'cpu')
ny = 2301
nx = 751
dx = 4.0
v = torch.from_file('vp.bin',
                    size=ny*nx).reshape(ny, nx).to(device)

n_shots = 1

n_sources_per_shot = 1
d_source = 20  # 20 * 4m = 80m
first_source = 10  # 10 * 4m = 40m
source_depth = 2  # 2 * 4m = 8m

n_receivers_per_shot = 384
d_receiver = 6  # 6 * 4m = 24m
first_receiver = 0  # 0 * 4m = 0m
receiver_depth = 2  # 2 * 4m = 8m

freq = 25
nt = 750
dt = 0.004
peak_time = 1.5 / freq

# source_locations
source_locations = torch.zeros(n_shots, n_sources_per_shot, 2,
                               dtype=torch.long, device=device)
source_locations[..., 1] = source_depth
source_locations[:, 0, 0] = (torch.arange(n_shots) * d_source +
                             first_source)

# receiver_locations
receiver_locations = torch.zeros(n_shots, n_receivers_per_shot, 2,
                                 dtype=torch.long, device=device)
receiver_locations[..., 1] = receiver_depth
receiver_locations[:, :, 0] = (
    (torch.arange(n_receivers_per_shot) * d_receiver +
     first_receiver)
    .repeat(n_shots, 1)
)

# source_amplitudes
source_amplitudes = (
    deepwave.wavelets.ricker(freq, nt, dt, peak_time)
    .repeat(n_shots, n_sources_per_shot, 1)
    .to(device)
)

n_shots = 1

n_sources_per_shot = 1
d_source = 20  # 20 * 4m = 80m
first_source = 10  # 10 * 4m = 40m
source_depth = 2  # 2 * 4m = 8m

n_receivers_per_shot = 384
d_receiver = 6  # 6 * 4m = 24m
first_receiver = 0  # 0 * 4m = 0m
receiver_depth = 2  # 2 * 4m = 8m

freq = 25
nt = 750
dt = 0.004
peak_time = 1.5 / freq

# source_locations
source_locations = torch.zeros(n_shots, n_sources_per_shot, 2,
                               dtype=torch.long, device=device)
source_locations[..., 1] = source_depth
source_locations[:, 0, 0] = (torch.arange(n_shots) * d_source +
                             first_source)

# receiver_locations
receiver_locations = torch.zeros(n_shots, n_receivers_per_shot, 2,
                                 dtype=torch.long, device=device)
receiver_locations[..., 1] = receiver_depth
receiver_locations[:, :, 0] = (
    (torch.arange(n_receivers_per_shot) * d_receiver +
     first_receiver)
    .repeat(n_shots, 1)
)

# source_amplitudes
source_amplitudes = (
    deepwave.wavelets.ricker(freq, nt, dt, peak_time)
    .repeat(n_shots, n_sources_per_shot, 1)
    .to(device)
)

def wrap(v, chunk, wavefield_0, wavefield_m1, psiy_m1, psix_m1,
         zetay_m1, zetax_m1):
    return scalar(
                  v, dx, dt,
                  source_amplitudes=chunk,
                  source_locations=source_locations,
                  receiver_locations=receiver_locations,
                  pml_freq=freq,
                  accuracy=8,
                  wavefield_0=wavefield_0,
                  wavefield_m1=wavefield_m1,
                  psiy_m1=psiy_m1,
                  psix_m1=psix_m1,
                  zetay_m1=zetay_m1,
                  zetax_m1=zetax_m1,
              )


n_epochs = 1
n_segments = 10
v.requires_grad_()
# Setup optimiser to perform inversion
optimiser = torch.optim.SGD([v], lr=1e9, momentum=0.9)
loss_fn = torch.nn.MSELoss()
for epoch in range(n_epochs):
    pml_width = 50
    wavefield_size = [n_shots, ny + 2 * pml_width,
                          nx + 2 * pml_width]
    wavefield_0 = torch.zeros(*wavefield_size, device=device)
    wavefield_m1 = torch.zeros(*wavefield_size, device=device)
    psiy_m1 = torch.zeros(*wavefield_size, device=device)
    psix_m1 = torch.zeros(*wavefield_size, device=device)
    zetay_m1 = torch.zeros(*wavefield_size, device=device)
    zetax_m1 = torch.zeros(*wavefield_size, device=device)
    optimiser.zero_grad()
    receiver_amplitudes2 = torch.zeros(n_shots,
                                          n_receivers_per_shot, nt,
                                          device=device)
    k = 0
    for i, chunk in enumerate(torch.chunk(source_amplitudes,
                                              n_segments,
                                              dim=-1)):
        if i == n_segments - 1:
                (wavefield_0, wavefield_m1, psiy_m1, psix_m1,
                zetay_m1, zetax_m1, receiver_amplitudes_chunk) = \
                wrap(v, chunk, wavefield_0,
                         wavefield_m1,
                         psiy_m1, psix_m1,
                         zetay_m1, zetax_m1)
        else:
            (wavefield_0, wavefield_m1, psiy_m1, psix_m1,
                 zetay_m1, zetax_m1, receiver_amplitudes_chunk) = \
                 torch.utils.checkpoint.checkpoint(wrap, v, chunk,
                                                   wavefield_0,
                                                   wavefield_m1,
                                                   psiy_m1, psix_m1,
                                                   zetay_m1,
                                                   zetax_m1)
        receiver_amplitudes2[..., k:k+chunk.shape[-1]] = \
                receiver_amplitudes_chunk
        k += chunk.shape[-1]

receiver_amplitudes = out[-1]
vmin, vmax = torch.quantile(receiver_amplitudes[0],
                            torch.tensor([0.05, 0.95]).to(device))
_, ax = plt.subplots(1, 2, figsize=(10.5, 7), sharey=True)
ax[0].imshow(receiver_amplitudes1[0].cpu().T, aspect='auto',
             cmap='gray', vmin=vmin, vmax=vmax)
ax[1].imshow(receiver_amplitudes2[0].detach().cpu().T, aspect='auto',
             cmap='gray', vmin=vmin, vmax=vmax)
ax[0].set_xlabel("Channel")
ax[0].set_ylabel("Time Sample")
ax[1].set_xlabel("Shot")
plt.tight_layout()
